run-xmllint-format.py

```diff
@@ -0,0 +1,323 @@
+"""
+
+import argparse
+import codecs
+import difflib
+import fnmatch
+import io
+import errno
+import multiprocessing
+import os
+import signal
+import subprocess
+import sys
+import traceback
+
+from functools import partial
+
+from subprocess import DEVNULL
+
+DEFAULT_EXTENSIONS = "xml"
+DEFAULT_XMLLINT_FORMAT_IGNORE = ".xmllint-format-ignore"
+
+
+class ExitStatus:
+    SUCCESS = 0
+    DIFF = 1
+    TROUBLE = 2
+
+
+def excludes_from_file(ignore_file):
+    excludes = []
+    with io.open(ignore_file, "r", encoding="utf-8") as f:
+        for line in f:
+            if line.startswith("#"):
+                # ignore comments
+                continue
+            pattern = line.rstrip()
+            if not pattern:
+                # allow empty lines
+                continue
+            excludes.append(pattern)
+    return excludes
+
+
+def list_files(files, recursive=False, extensions=None, exclude=None):
+    if extensions is None:
+        extensions = []
+    if exclude is None:
+        exclude = []
+
+    out = []
+    for file in files:
+        if recursive and os.path.isdir(file):
+            for dirpath, dnames, fnames in os.walk(file):
+                fpaths = [os.path.join(dirpath, fname) for fname in fnames]
+                for pattern in exclude:
+                    # os.walk() supports trimming down the dnames list
+                    # by modifying it in-place,
+                    # to avoid unnecessary directory listings.
+                    dnames[:] = [
+                        x
+                        for x in dnames
+                        if not fnmatch.fnmatch(os.path.join(dirpath, x), pattern)
+                    ]
+                    fpaths = [x for x in fpaths if not fnmatch.fnmatch(x, pattern)]
+                for f in fpaths:
+                    ext = os.path.splitext(f)[1][1:]
+                    if ext in extensions:
+                        out.append(f)
+        else:
+            out.append(file)
+    return out
+
+
+def make_diff(file, original, reformatted):
+    return list(
+        difflib.unified_diff(
+            original,
+            reformatted,
+            fromfile="{}\t(original)".format(file),
+            tofile="{}\t(reformatted)".format(file),
+            n=3,
+        )
+    )
+
+
+class DiffError(Exception):
+    def __init__(self, message, errs=None):
+        super(DiffError, self).__init__(message)
+        self.errs = errs or []
+
+
+class UnexpectedError(Exception):
+    def __init__(self, message, exc=None):
+        super(UnexpectedError, self).__init__(message)
+        self.formatted_traceback = traceback.format_exc()
+        self.exc = exc
+
+
+def run_xmllint_format_diff_wrapper(args, file):
+    try:
+        ret = run_xmllint_format_diff(args, file)
+        return ret
+    except DiffError:
+        raise
+    except Exception as e:
+        raise UnexpectedError("{}: {}: {}".format(file, e.__class__.__name__, e), e)
```

Comment:
Please consider f strings instead of .format. It's almost always clearer (the data being expanded appears right where it gets expanded).

Also please consider `raise from`:
```
except Exception as e
    raise UnexpectedError(f"run_xmllint_format_diff failed on {file}") from e
```
It gives the user more information and avoids the need to duplicate any information from the original exception.

Please consider `raise ... from ...` for all other instances where you also catch one exception and raise another (assuming the try/catch adds value).